sdk-utils/src/main/java/com/redhat/parodos/sdkutils/RetryExecutorService.java

```diff
@@ -37,48 +39,41 @@ public T submitWithRetry(Callable<T> task) {
 	 */
 	public T submitWithRetry(Callable<T> task, Runnable onSuccess, Runnable onFailure, long maxRetryTime,
 			long retryDelay) {
-		Future<T> future = executor.submit(() -> {
-			long startTime = System.currentTimeMillis();
-			long endTime = startTime + maxRetryTime;
+		CompletableFuture<T> future = new CompletableFuture<>();
+		long startTime = System.currentTimeMillis();
+		long endTime = startTime + maxRetryTime;
 
-			while (System.currentTimeMillis() < endTime) {
-				try {
-					T result = task.call();
-					onSuccess.run();
-					return result; // Success, no need to retry
-				}
-				catch (Exception e) {
-					// Task failed, invoke onFailure callback
-					onFailure.run();
-
-					// Sleep for the retry delay
-					try {
-						// FIXME: This is a blocking call, we should use a non-blocking
-						// sleep
-						Thread.sleep(retryDelay);
-					}
-					catch (InterruptedException ex) {
-						Thread.currentThread().interrupt();
-						return null; // Interrupted, exit the task
-					}
-				}
+		ScheduledFuture<?> scheduledFuture = scheduledExecutor.scheduleWithFixedDelay(() -> {
+			if (System.currentTimeMillis() >= endTime) {
+				future.completeExceptionally(new TimeoutException("Retry limit reached."));
+				return;
 			}
```

Comment:
Set a timeout in `future.get()` (i.e. `future.get(endTime, TimeUnit.MILLISECONDS);`  does it have equivalent behavior to this block of code?

Reply:
No, setting a timeout using `future.get(endTime, TimeUnit.MILLISECONDS)` does not provide an equivalent behavior to the previous block of code in the RetryExecutorService.

The `future.get(endTime, TimeUnit.MILLISECONDS)` method call sets a timeout for waiting to retrieve the result from the future. If the result is not available within the specified timeout period, a TimeoutException will be thrown. However, it does not cancel the ongoing execution of the task or stop further retry attempts.